### ISSUE TITLE
Multiple value on List-Unsubscribe header

### ISSUE THREAD

**Issue by LevryKurniawan** [mautic/mautic#6339]
What type of report is this:

| Q  | A
| ---| ---
| Bug report? | Y
| Feature request? | N
| Enhancement? | N

## Description:
I found multiple value on List-Unsubscribe header in every email sent via Mautic if using the Monitored Inbox feature. This happens to all emails sent directly, campaigns, broadcasts, or actions on the form.

![](https://user-images.githubusercontent.com/40631294/43300962-709daab2-918c-11e8-976f-5c4300c42d53.png)

## If a bug:

| Q   | A
| --- | ---
| Mautic version | 2.14.0
| PHP version | 5.6.36

## Steps to reproduce:
1. Complete all fields in Monitored Inbox Settings.
2. Perform the procedure to send emails to contacts in any way.
3. Check the header of email that the contact received.
 
## Log errors: 
Nothing.

**Comment by npracht** [comment]
Hello @levrykurniawan,

There is a template for creating issues and as far as i remember there is a Description section and also a How to reproduce section.
Feel free to complete ;)

**Comment by PauloMcGuinee** [comment]
Hi All,

I am having the same or similar issue after the latest stable upgrade.

To Replicate:

a. Select Queue mode for emails. Mine is set at 300/HOUR
b. Cron Jobs set to send campaigns each hour.

1. In Channels|Emails create a New  Segment Email campaign. 
2. Code Template
3. Save and Publish for sending

Now since the latest upgrade to 2.14 - only about 10 emails send each hour.

By looking in the Mautic directory and navigating to Spool - I can see that each individual email in the queue is get progressively bigger in size from a few hundred kilobytes to 1.8mb (same email content)

So I downloaded a few as samples and can see that the header is "growing" on each email until it is absolutely massive. Much like the OP above.

**Comment by alanhartless** [comment]
Does this only happen when email is queued to the filesystem and processed by a cron job?

**Comment by PauloMcGuinee** [comment]
I can't easily answer that as I do not have a system configured any other way. I only have a live version so am reluctant to run any tests on it and risk firing emails to real people! However, it would be a real effort to roll Mautic back to the previous version so I am a bit stuck right now with only a few emails going out an hour and most being rejected for long headers. 

Original Posterare you able to help answer this question - or can anyone else replicate the issue?

**Comment by PauloMcGuinee** [comment]
My best guess is that instead of adding the next unsubscribe link in a loop to each queued message, Mautic has started appending the next one onto the string. Therefore each message is getting longer and longer. I have 5000 emails in this particular campaign so the last email to be queued probably has 5000 unsubscribe links in the header! If the same code generates the emails regardless of the process of sending selected then I expect this bug will show itself in all New Segment Email campaigns.

**Comment by PauloMcGuinee** [comment]
One other piece of information is that I always clone a previous campaign and haven't started a fresh campaign for a long time. So I am cloning a campaign sent in a previous version and then publishing it in the latest release version. Might not be relevant - but worth a mention.

**Comment by PauloMcGuinee** [comment]
Has anyone been able to replicate this? It is a real issue as I can no longer send campaigns and have had to manually clear the spool and just put everything on hold. I'm keen to see if there is a workaround or a fix before going through the long process of restoring an older version.

**Comment by PauloMcGuinee** [comment]
@alanhartless did you get chance to look into this particular issue. I would assume more people would be experiencing it - unless it is unique to the particular set-up I have. Any help would be appreciated.

**Comment by PauloMcGuinee** [comment]
@levrykurniawan  did you find a solution?

**Comment by levrykurniawan** [comment]
Hello @npracht,

I have updated Description and How to reproduce section in the post above.

I don't use queues like @PauloMcGuinee. But I think this problem is closely related. I also found the same thing in an email sent by my friend who also uses Mautic.

The temporary solution to this problem is not to use the Monitored Inbox feature.

**Comment by levrykurniawan** [comment]
Hello @npracht,

I have updated Description and Steps to reproduce section in the post above.

I don't use queues like @PauloMcGuinee. But I think this issue is closely related. I also found the same thing in an email sent by my friend who also uses Mautic.

The temporary solution to this issue is not to use the Monitored Inbox feature.

**Comment by npracht** [comment]
Hm, i guess something is going wrong here.
The unsubscribe link if you click on it, you arrive on the unsubscribe page of whom? The monitored inbox or the people to whom you addressed the email?

**Comment by levrykurniawan** [comment]
I guess there is no problem with unsubscribe link, it works properly. It will go to the unsubscribe page as usual, and the person who clicks it will be marked as Do Not Contact.

**Comment by levrykurniawan** [comment]
In the broadcast email, the value will be even more.

![img_20180808_160128](https://user-images.githubusercontent.com/40631294/43827980-74c0adb8-9b25-11e8-8f1c-8d2100908490.JPG)

**Comment by PauloMcGuinee** [comment]
Yes, this is my issue. It is crazy. @levrykurniawan - "The temporary solution to this issue is not to use the Monitored Inbox feature." What is the method to disable this for now?

**Comment by levrykurniawan** [comment]
Go to **Settings » Configuration » Email Settings**. Leave this field blank.

![cuplikan layar dari 2018-08-08 16-23-57](https://user-images.githubusercontent.com/40631294/43828746-90609b12-9b27-11e8-8e78-7275f9d74862.png)

**Comment by StudioMaX** [comment]
So, the reason is that property `Mautic\EmailBundle\Helper\MailHelper::$headers` is set once while initialization of the batch, and adds "custom headers" here (only once). However, after this, during the foreach of each contact, these headers are not cleared.
In fact, there are two issues:
1. The same method of unsubscription (unsubscribe@email and URL) is duplicated several times in each prepared email. Just look at `MailHelper::getCustomHeaders()` - there are no checks that needed URL is already existed in the variable `$headers['List-Unsubscribe']`. Although this method is called repeatedly, again and again adding duplicates, even when it is not needed. For example `Mautic\LeadBundle\EventListener\EventListener::onEmailGenerate()` calls `$event->getTextHeaders()->getCustomHeaders()`. Exactly the same issue with duplicating unsubscribe email addresses, but already in another object `Mautic\EmailBundle\EventListener::onEmailSend()`.
2. Another problem, much more serious - the header "List-Unsubscribe" contains URLs/emails belonging to those contacts who were processed in the same queue request. This is because the `MailHelper::$headers` are not reset when each user is iterated.

I don't know the architecture of Mautic and the correct way to fix this bug. For myself, I fixed it this way:
https://github.com/mautic/mautic/blob/729fa25e276d233c495c6de743279cf9078642e8/app/bundles/EmailBundle/EventListener/ProcessUnsubscribeSubscriber.php#L97
replace to
```php
            if ($existing) {
                if (strpos($existing, $unsubscribeEmail) === false) {
                    $updatedHeader = $unsubscribeEmail . ', ' . $existing;
                } else {
                    $updatedHeader = $existing;
                }
            } else {
                $updatedHeader = $unsubscribeEmail;
            }
```
---
https://github.com/mautic/mautic/blob/729fa25e276d233c495c6de743279cf9078642e8/app/bundles/EmailBundle/Helper/MailHelper.php#L1474-L1475
replace to
```php
                if (strpos($headers['List-Unsubscribe'], $listUnsubscribeHeader) === false) {
                    // Ensure Mautic's is always part of this header
                    $headers['List-Unsubscribe'] .= ',' . $listUnsubscribeHeader;
                }
```
---
https://github.com/mautic/mautic/blob/729fa25e276d233c495c6de743279cf9078642e8/app/bundles/EmailBundle/Helper/MailHelper.php#L2105-L2121
replace to (or just copy the last `if` block at the end of the method)
```php
    private function setMessageHeaders()
    {
        $headers = $this->getCustomHeaders();

        // Set custom headers
        if (!empty($headers)) {
            $messageHeaders = $this->message->getHeaders();
            foreach ($headers as $headerKey => $headerValue) {
                if ($messageHeaders->has($headerKey)) {
                    $header = $messageHeaders->get($headerKey);
                    $header->setFieldBodyModel($headerValue);
                } else {
                    $messageHeaders->addTextHeader($headerKey, $headerValue);
                }
            }
        }

        if (array_key_exists('List-Unsubscribe', $headers)) {
            unset($headers['List-Unsubscribe']);
            $this->setCustomHeaders($headers, false);
        }
    }
```

Note: If you have added `List-Unsubscribe` via "custom headers", it will be lost.

**Comment by KalleVuorjoki** [comment]
I can confirm this. `List-Unsubscribe` got duplicated, after applying @StudioMaX's changes and manually removing `spool` directory, the `List-Unsubscribe` header got fixed and being able to send messages.

**Comment by GabriGreese** [comment]
I can confirm too.

**Comment by tomingoglia** [comment]
I'm running the latest version (2.14.2) and this is still a problem.

**Comment by AlbanL74fr** [comment]
Me too, I still have the problem. And the log says : `Expected response code 250 but got code "554", with message "554 Transaction failed: Header too long: 'List-Unsubscribe'.`

**Comment by AlbanL74fr** [comment]
@escopecz Do you think the chunking options (--batch) of `mautic:broadcasts:send` help mitigate this issue?

**Comment by tomingoglia** [comment]
Hi @AlbanL74fr ...I think I went down the same path that you did with no success. What did ultimately work for me was the code fixes that @StudioMaX provided above (awesome and thank you for sharing!). They were simple updates that only took me a few minutes to apply. It would be nice if someone on the team would acknowledge this as an issue. The problem has been clearly defined, it is easily reproducible, and there is a proposed fix. It has clearly been around for a while and effects anyone who is sending an email out...which is likely everyone using this system. I'm not sure if it is just an SES problem. My guess is that it is a problem no matter what provider you are using and maybe only certain providers kick back an error when the header gets too long. Clearly the header is being built out incorrectly. A major downside here is that to users, this may appear to be working properly. Unless you go and look at the spool folder, you might think that your system is sending out all of the email. It's unfortunate that new users to Mautic need to struggle through this fundamental issue with little attention from the team. I spent many hours on this. The upside is that I learned a lot about how Mautic works in the process.

**Comment by tomingoglia** [comment]
@rafaelpatro, you can install Mautic and then manually modify the two php pages as prescribed by @StudioMaX. It looks a little intimidating if you don't write code, but it is quite simple to replace the code in question. Using a text editor with line numbers will be helpful or you can download [Visual Studio Code](https://code.visualstudio.com/) which is a very nice cross platform editor for free. If you struggle, reply and I can attach my edited files.

Hopefully @alanhartless will be able to have a look soon.

**Comment by rafaelpatro** [comment]
Thanks @tomingoglia 
Yes I've updated the two files above. But it did not take effect in my running E-mail.
I'll try create a new segmented E-mail.

**Comment by rafaelpatro** [comment]
Files updated:
_mautic/app/bundles/EmailBundle/Helper/MailHelper.php_
_mautic/app/bundles/EmailBundle/EventListener/ProcessUnsubscribeSubscriber.php_

And segmented E-mails are not working...
![image](https://user-images.githubusercontent.com/13813964/48673177-ec905780-eb25-11e8-888f-5fe22811b704.png)

**Comment by rafaelpatro** [comment]
I guess more information can help in a solution...

E-mail Service: Amazon SES
Monitored Inbox Settings: Disabled (all fields empty)
Files updated: _MailHelper.php_ and _ProcessUnsubscribeSubscriber.php_ as mentioned above.
Segmented Test E-mail: Working
Mautic version: 2.14.2

**Comment by rafaelpatro** [comment]
I'm trying to send using browser too and progress shows that its sending. But no one e-mail was sent.
![image](https://user-images.githubusercontent.com/13813964/48674925-143ceb00-eb39-11e8-838a-385825266b3f.png)

**Comment by tomingoglia** [comment]
@rafaelpatro are there email in the app/spool/default folder? If so, you might want to delete all of those and try again. Email get queued into that folder and then sent out via smtp by another process. Be sure that both cron jobs are running. I cleaned up that folder after fixing the above files. The files that are in that folder will not be fixed by the above changes because they were the output of the bad code.

Sorry for the delayed response. I hope that helps.

**Comment by bscx** [comment]
@npracht Hey Norman, This is a bug that prevents mautic from sending emails that contain List-Unsubscribe headers. Has this issue been resolved in mautic 2.15.0? If no, is there any reason why you wouldn't fix such a critical bug? Thanks for your effort and your quick response.

**Comment by Mike-Benoit** [comment]
We are seeing this issue too, however the emails are being sent out, but the vast majority of them just get bounced back on the receiving email server with some error like:

Diagnostic-Code: smtp; 554 5.6.0 Invalid message content

**Comment by bscx** [comment]
> Hello @bscx not fixed as far as i know. Why it is not done yet, because none from the community took time to invest in it. Feel free to contribute :)

Hello @npracht Thanks for your fast response and I really appreciate your effort. However, please help me to understand this: @StudioMaX suggested a fix on 11 Aug 2018 and there was no "official" acknowledgement of his work although it perfectly fixes the problem.

@Mike-Benoit now confirms that this crucial bug is still in the code base. This again leads to my conclusion that a huge number of systems must be affected that renders mautic virtually useless.

What exactly is needed to contribute? As far as I can see the bug is not even acknowledged although it's reproducible.

**Comment by escopecz** [comment]
@StudioMaX could you send a pull request?

@bscx The code changes proposals to open source projects are made with pull requests. I didn't even know that such comment with proposed fix existed. It would be merged to core by now otherwise. Someone just have to take it and make the pull request. 

The issue gets acknowledged as a bug by the `bug` label and this issue is labeled as such. Would you suggest other way to do it?

**Comment by bscx** [comment]
@escopecz Hello John, Thanks for your reply! I'm aware of pull requests and I would accept the label "bug" as acknowledgment of the bug. However, @npracht set it from "pending feedback" to "bug" and although there was a proposal by @StudioMaX (not via pull request though), it seems like mautic core developers were still waiting on feedback.
This was maybe a misunderstanding and I'll be more than happy to help you out with this.
The label "bug" was misleading to me as I'm used to check for "acknowledged by core developers" or "acknowledged bug". Appreciate your effort!

**Comment by afonseca08** [comment]
I can report the issue persists in version 2.15.0 and applying the code changes by @StudioMaX manually fixes it for me as well. Hopefully this can make it into the next release.

**Comment by KalleVuorjoki** [comment]
I took the liberty to create a PR in behalf of StudioMax. Hopefully this get committed soon.

**Comment by bulgin** [comment]
I can confirm applying the fix suggested by @StudioMaX  does not work for me in 2.15.0.

**Comment by bscx** [comment]
> I can confirm applying the fix suggested by @StudioMaX does not work for me in 2.15.0.

Are you sure you have applied the patch from @StudioMaX correctly? I have applied it to our version 2.15.0 and it seems to be working like a charm.

**Comment by vincnetas** [comment]
Another workaround if : 

- you already have lots of files in spool folder
- not sure what was already sent
- don't want to send duplicate files

Modify : app/bundles/EmailBundle/Command/ProcessEmailQueueCommand.php

```
// Workaround for  https://github.com/mautic/mautic/issues/6339
// Remove all 'list-unsubscribe' headers from spooled files 
$finder = Finder::create()->in($spoolPath)->name('*.{message}');
foreach ($finder as $file) {
    $file = $file->getRealPath();

    $message = unserialize(file_get_contents($file));
    if ($message->getHeaders()->has('list-unsubscribe')) {
        $message->getHeaders()->removeAll('list-unsubscribe');
        file_put_contents($file, serialize($message));
    }
}
// end workaround

```


This can be inserted after :

```
$spoolPath = $container->getParameter('mautic.mailer_spool_path');
if (file_exists($spoolPath)) {

```

**Comment by emackaran** [comment]
Hi have the same issue and i am trying to apply the fix suggested by @StudioMaX , however my issue is I cannot find the php files to modify. I am not a programmer like you guys and I cannot find the php files. 

I am using mautic installation on aws by bitnami, search through using ssh connection but cannot locate the files to edit any help will be appreciated

**Comment by djeraseit** [comment]
@vincnetas does your workaround allow keeping the monitored inbox settings?

**Comment by vincnetas** [comment]
> @vincnetas does your workaround allow keeping the monitored inbox settings?

As far as i know this workaround had no efect on monitored indox, or else my collegues would have complained by now.

**Comment by techzotezo** [comment]
@StudioMaX  after adding your code in 
mautic/app/bundles/EmailBundle/Helper/MailHelper.php it is showing this error in log

[2019-02-28 13:04:04] mautic.CRITICAL: Uncaught PHP Exception Symfony\Component\Debug\Exception\FatalThrowableError: "Parse error: syntax error, unexpected end of file" at mautic/app/bundles/EmailBundle/Helper/MailHelper.php line 1655 {"exception":"[object] (Symfony\\Component\\Debug\\Exception\\FatalThrowableError(code: 0): Parse error: syntax error, unexpected end of file at mautic/app/bundles/EmailBundle/Helper/MailHelper.php:1655)"} [] []

**Comment by nutriepaul** [comment]
is there a pull request I can pull in?  

We have over 250,000 emails and its taking a while to build the emails. i've implemented the spool fix, but i'd love to prevent these emails from getting built like this.

**Comment by KalleVuorjoki** [comment]
@nutriepaul yes #7110

**Comment by pradeepm** [comment]
@vincnetas Could you please let me know if we replace the existing code with the snippet you've provided or just add it there? If we are a replacing, please kindly mention up to which line that we replace it with your snippet. Thanks

**Comment by KalleVuorjoki** [comment]
PR linked here is not merged in 2.15.1, so please report if the issue still exists in 2.15.1.

**Comment by luisenrike** [comment]
I'm still having the issue in 2.15.1

Also, because of this, my disk space is at 900gb

**Comment by npracht** [comment]
Please @luisenrike test #7110

**Comment by AkkuFresh** [comment]
I can confirm, I also have this issue with Mautic 2.15.0 ( PHP 7.2 )

**Comment by npracht** [comment]
Finally @LevryKurniawan @PauloMcGuinee @AkkuFresh does #7110 fixes the issue ?

**Comment by LevryKurniawan** [comment]
I didn't test that PR because I fixed this issue by disabling **Monitored Inbox** as I explained above.

**Comment by AkkuFresh** [comment]
@npracht  Thank you, I have just tested this fix in our production environment, and it works well, as it was expected.

**Comment by mautibot** [comment]
This issue has been mentioned on **Mautic Community Forums**. There might be relevant details there:

https://forum.mautic.org/t/list-unsubscribe-header/16945/8

**Comment by adirbnology** [comment]
hi. this issue occurs again on 3.2.5.
List-Unsubscribe containing all contacts in queue request.
this is critical issue. 

tested on amazon api.

**Comment by RCheesley** [comment]
Please raise a new issue and link to this old issue!
